[BEAM-10891] Dockerfile for development container (#12837)
apache/beam: website/www/site/content/en/contribute/_index.md

gLinux users should configure their machines for sudoless Docker.

> Alternatively, you can use the [Dockerfile](https://github.com/apache/beam/blob/master/Dockerfile) to setup a container meeting the requirements above, and mounting the directory that contains the Beam repo to the container. e.g.:

Reviewer: ```suggestion
Alternatively, you can use the [Dockerfile](https://github.com/apache/beam/blob/master/Dockerfile) to setup a container meeting the requirements above, and mount your clone of the Beam repo to the container. e.g.:
```

```shell script
docker run -it --mount type=bind,source=/Users/jsmith/Desktop/beam,target=/workspaces/beam <CONTAINER_ID> bash

> ```

Reviewer: It would be nice if this also had directions for building the container and getting the id, something like:
```sh
docker build --rm -t beam-dev ${BEAM_DIR}
set image_id=$(docker images -q beam-dev)
docker run -it --mount type=bind,source=${BEAM_DIR},target=/workspaces/beam ${image_id} bash
```

Reviewer 2: I would prefer a script similar (same) as what is present in the Hadoop project:
https://github.com/apache/hadoop/blob/trunk/start-build-env.sh
It makes starting/building the development environment for the developers a lot easier as it also mounts the working directory and has fixes and workarounds for various situations (like selinux, MacOS, etc.)

Author: I've never seen that before, but I can give it a shot. How hard can it be? 🙀

Author: @nielsbasjes I was going off how Airflow does it:
https://github.com/apache/airflow/blob/master/Dockerfile

Reviewer: Thanks @nielsbasjes that's a great example. Before pursuing a script like that though I think we should make sure there's some way we can make it possible to build and run docker containers from within this dev container. Do we just need to add some binding for `unix:///var/run/docker.sock`?

This [blog post](https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/) has a suggestion:
```
docker run -v /var/run/docker.sock:/var/run/docker.sock ...
```

If that makes it possible to run `./gradlew :sdks:python:container:py37:docker` and we can roll it into a `start-build-env.sh` script I'd be in favor

Author: I can look into it as well!

Reviewer 2: This /var/run/docker.sock trick seems like something other projects would like also.
I do wonder how portable this is. I expect this to work under Linux. But how does that work on a Mac or even something like Windows?
I also wonder is this will work on all Linux systems. For one I have read that CentOS 8 no longer installs docker in favor of Podman/Buildah (but there is a podman-docker package so you can still have the same commands). 

Looking at this effort I propose to have multiple goals to achieve:

1. Have the docker image that can do "most" of the build (i.e. skip the parts that need docker). 
    Seems like this is already there now.
1. Include the script I mentioned so developers can already use this for a lot of the work. 
    Should be an easy addition.
1. Figure out how to run docker commands inside this environment and add that as a later enhancement.
    I expect this to be a major additional effort because of the variations in underlying implementations.
    A first version can simply be `if /var/run/docker.sock then "enable this" else "simply don't" `

Author: I used 

```
docker run -v /var/run/docker.sock:/var/run/docker.sock ...
```

on my Mac and it worked! I see on your PR, you used a different method. Would that work for all OSes?
